File: neetcode150/backtracking/subsets.py
#using backtracking
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res=[]
        subset = []
        n = len(nums)

        def dfs(i):

            if i >= n:
                res.append(subset.copy())
                return 
            subset.append(nums[i])
            dfs(i+1)
            subset.pop()
            dfs(i+1)
        
        dfs(0)
        return res

  '''not using backtracking
  class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = [[]]

        for n in nums:
            if [n] not in res:
                res.append([n])
            
            for l in res:
                if l == [n]:
                    break
                new = l + [n]
                if new not in res:
                    res.append(new)
        
        return res
  '''

File: neetcode150/graphs/subsets.cpp
class Solution {
    vector<vector<int>> output;
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> set;
        recursion(0, set, nums, nums.size());
        return output;
    }
private:
    void recursion (int i, vector<int>& set, vector<int>& nums, int n){
        if (i==n){
            output.push_back(set);
            return;
        }

        set.push_back(nums[i]);
        recursion(i+1, set, nums, n);
        set.pop_back();
        recursion(i+1, set, nums, n);
    }
};
